### PR TITLE
Feature: subscription engine integration

### LINKED ISSUE #832: Integrate subscription engine

### PR DESCRIPTION
**Pull request by iceseer**
### Linked issue

Closes #832

### Description of the Change
New subscription system in iroha

### CODE REVIEW

File: irohad/ametsuchi/impl/storage_impl.cpp [hunk 1]
```diff
@@ -61,7 +62,7 @@ namespace iroha {
         : block_store_(std::move(block_store)),
           pool_wrapper_(std::move(pool_wrapper)),
           connection_(pool_wrapper_->connection_pool_),
-          notifier_(notifier_lifetime_),
+          // notifier_(notifier_lifetime_),
```
Comment (kuvaldini) [inline irohad/ametsuchi/impl/storage_impl.cpp:65]: remove if not required

**Review by kuvaldini — CHANGES_REQUESTED** [review]

File: irohad/ordering/impl/on_demand_ordering_gate.hpp [hunk 2]
```diff
@@ -93,17 +97,30 @@ namespace iroha {
...
+      std::shared_ptr<ProcessedHashesSubscriberType>
+          processed_hashes_subscription_;
+      std::shared_ptr<RoundSwitchSubscriberType> round_switch_subscription_;
+
       std::shared_timed_mutex stop_mutex_;
       bool stop_requested_{false};
 
-      rxcpp::composite_subscription proposal_notifier_lifetime_;
-      rxcpp::subjects::subject<network::OrderingEvent> proposal_notifier_;
+      // rxcpp::composite_subscription proposal_notifier_lifetime_;
```
Comment (kuvaldini) [inline irohad/ordering/impl/on_demand_ordering_gate.hpp:121]: rm

File: irohad/ordering/impl/on_demand_ordering_gate.hpp [hunk 3]
```diff
@@ -93,17 +97,30 @@ namespace iroha {
       size_t transaction_limit_;
       std::shared_ptr<OnDemandOrderingService> ordering_service_;
       std::unique_ptr<transport::OdOsNotification> network_client_;
-      rxcpp::composite_subscription processed_tx_hashes_subscription_;
-      rxcpp::composite_subscription round_switch_subscription_;
+      // rxcpp::composite_subscription processed_tx_hashes_subscription_;
```
Comment (kuvaldini) [inline irohad/ordering/impl/on_demand_ordering_gate.hpp:100]: rm

File: irohad/ordering/impl/on_demand_ordering_gate.hpp [hunk 4]
```diff
@@ -63,7 +65,9 @@ namespace iroha {
           std::shared_ptr<shared_model::interface::TransactionBatch> batch)
           override;
 
-      rxcpp::observable<network::OrderingEvent> onProposal() override;
+      // rxcpp::observable<network::OrderingEvent> onProposal() override;
```
Comment (kuvaldini) [inline irohad/ordering/impl/on_demand_ordering_gate.hpp:68]: rm

File: irohad/ordering/impl/on_demand_ordering_gate.hpp [hunk 5]
```diff
@@ -46,10 +48,10 @@ namespace iroha {
       OnDemandOrderingGate(
           std::shared_ptr<OnDemandOrderingService> ordering_service,
           std::unique_ptr<transport::OdOsNotification> network_client,
-          rxcpp::observable<
+          /*rxcpp::observable<
               std::shared_ptr<const cache::OrderingGateCache::HashesSetType>>
               processed_tx_hashes,
-          rxcpp::observable<RoundSwitch> round_switch_events,
+          rxcpp::observable<RoundSwitch> round_switch_events,*/
```
Comment (kuvaldini) [inline irohad/ordering/impl/on_demand_ordering_gate.hpp:54]: rm

File: irohad/ordering/impl/on_demand_ordering_gate.hpp [hunk 6]
```diff
@@ -46,10 +48,10 @@ namespace iroha {
       OnDemandOrderingGate(
           std::shared_ptr<OnDemandOrderingService> ordering_service,
           std::unique_ptr<transport::OdOsNotification> network_client,
-          rxcpp::observable<
+          /*rxcpp::observable<
```
Comment (kuvaldini) [inline irohad/ordering/impl/on_demand_ordering_gate.hpp:51]: rm

**Review by kuvaldini — CHANGES_REQUESTED** [review]

File: irohad/main/application.hpp [hunk 7]
```diff
@@ -370,6 +372,33 @@ class Irohad {
...
+  std::shared_ptr<iroha::subscription::SubscriberImpl<
+      iroha::EventTypes,
+      iroha::SubscriptionDispatcher,
+      bool,
+      iroha::synchronizer::SynchronizationEvent>>
+      syncSubscription;
+
+  using OnProposalSubscription =
+      iroha::subscription::SubscriberImpl<iroha::EventTypes,
+                                          iroha::SubscriptionDispatcher,
```
Comment (kuvaldini) [inline irohad/main/application.hpp:385]: I did not see any SubscriberImpl with tparams other than <iroha::EventTypes,iroha::SubscriptionDispatcher,...>

Lets `using` it as BasicSubsciber<Receiver,NotificationArgs>

**Review by kuvaldini — CHANGES_REQUESTED** [review]

File: irohad/torii/processor/transaction_processor_impl.hpp [hunk 8]
```diff
@@ -67,6 +68,33 @@ namespace iroha {
 
       logger::LoggerPtr log_;
 
+      using VerifiedProposalSubscription =
+          subscription::SubscriberImpl<EventTypes,
+                                       SubscriptionDispatcher,
+                                       bool,
+                                       simulator::VerifiedProposalCreatorEvent>;
+      using BlockSubscription = subscription::SubscriberImpl<
```
Comment (kuvaldini) [inline irohad/torii/processor/transaction_processor_impl.hpp:76]: This code is duplicated at least in 4 places. Please unify.

File: irohad/consensus/yac/yac.hpp [hunk 9]
```diff
@@ -18,6 +18,7 @@
 #include "consensus/yac/outcome_messages.hpp"  // because messages passed by value
 #include "consensus/yac/storage/yac_vote_storage.hpp"  // for VoteStorage
 #include "logger/logger_fwd.hpp"
+#include "main/subscription.hpp"
 
 #include <rxcpp/operators/rx-observe_on.hpp>
```
Comment (kuvaldini) [inline irohad/consensus/yac/yac.hpp:23]: remove this rx-cpp includes and from vcpkg

File: irohad/multi_sig_transactions/impl/mst_processor_impl.cpp [hunk 10]
```diff
@@ -12,6 +12,7 @@
 #include <rxcpp/operators/rx-map.hpp>
 #include <rxcpp/operators/rx-take.hpp>
```
Comment (kuvaldini) [inline irohad/multi_sig_transactions/impl/mst_processor_impl.cpp:13]: rm

File: irohad/multi_sig_transactions/storage/mst_storage_impl.hpp [hunk 11]
```diff
@@ -11,6 +11,7 @@
 
 #include <rxcpp/rx-lite.hpp>
```
Comment (kuvaldini) [inline irohad/multi_sig_transactions/storage/mst_storage_impl.hpp:12]: rm

File: test/framework/test_subscriber.hpp [hunk 12]
```diff
@@ -10,6 +10,9 @@
 #include <memory>
 #include <rxcpp/rx-lite.hpp>
```
Comment (kuvaldini) [inline test/framework/test_subscriber.hpp:11]: rm

File: irohad/ordering/impl/on_demand_connection_manager.hpp [hunk 13]
```diff
@@ -13,6 +13,7 @@
 
 #include <rxcpp/rx-lite.hpp>
```
Comment (kuvaldini) [inline irohad/ordering/impl/on_demand_connection_manager.hpp:14]: rm

File: irohad/pending_txs_storage/impl/pending_txs_storage_impl.hpp [hunk 14]
```diff
@@ -20,6 +20,7 @@
 #include <rxcpp/rx-lite.hpp>
```
Comment (kuvaldini) [inline irohad/pending_txs_storage/impl/pending_txs_storage_impl.hpp:20]: rm

File: irohad/main/application.cpp [hunk 15]
```diff
@@ -128,13 +129,37 @@ Irohad::Irohad(
       yac_init(std::make_unique<iroha::consensus::yac::YacInit>()),
       consensus_gate_objects(consensus_gate_objects_lifetime),
       log_manager_(std::move(logger_manager)),
-      log_(log_manager_->getLogger()) {
+      log_(log_manager_->getLogger()),
+      on_block_subscription_(std::make_shared<OnBlockSubscription>(
```
Comment (kuvaldini) [inline irohad/main/application.cpp:133]: Please use the same name everywhere `OnBlockSubscription` or `BlockSubscription`

File: irohad/ametsuchi/impl/storage_impl.cpp (hunk 1 above)
Comment (iceseer) [reply irohad/ametsuchi/impl/storage_impl.cpp:65]: https://imgur.com/3l4KntA.png

File: irohad/ordering/impl/on_demand_ordering_gate.hpp (hunk 2 above)
Comment (iceseer) [reply irohad/ordering/impl/on_demand_ordering_gate.hpp:121]: https://imgur.com/3l4KntA.png

File: irohad/ordering/impl/on_demand_ordering_gate.hpp (hunk 3 above)
Comment (iceseer) [reply irohad/ordering/impl/on_demand_ordering_gate.hpp:100]: https://imgur.com/3l4KntA.png

File: irohad/ordering/impl/on_demand_ordering_gate.hpp (hunk 4 above)
Comment (iceseer) [reply irohad/ordering/impl/on_demand_ordering_gate.hpp:68]: https://imgur.com/3l4KntA.png

File: irohad/ordering/impl/on_demand_ordering_gate.hpp (hunk 5 above)
Comment (iceseer) [reply irohad/ordering/impl/on_demand_ordering_gate.hpp:54]: https://imgur.com/3l4KntA.png

File: irohad/main/impl/on_demand_ordering_init.hpp [hunk 16]
```diff
@@ -173,11 +174,28 @@ namespace iroha {
       std::shared_ptr<grpc::Service> service;
 
       /// commit notifier from peer communication service
-      rxcpp::subjects::subject<synchronizer::SynchronizationEvent>
-          sync_event_notifier;
-      rxcpp::subjects::subject<
+      /*rxcpp::subjects::subject<synchronizer::SynchronizationEvent>
```
[the remark this answers is not in the document]
Comment (iceseer) [reply irohad/main/impl/on_demand_ordering_init.hpp:177]: https://imgur.com/3l4KntA.png

File: irohad/main/impl/pending_transaction_storage_init.cpp [hunk 17]
```diff
@@ -15,36 +15,37 @@
 using namespace iroha;
 
 PendingTransactionStorageInit::PendingTransactionStorageInit()
-    : updated_batches(pending_storage_lifetime),
-      prepared_batch(pending_storage_lifetime),
-      expired_batch(pending_storage_lifetime),
-      prepared_txs(pending_storage_lifetime) {}
+/*: updated_batches(pending_storage_lifetime),
```
[the remark this answers is not in the document]
Comment (iceseer) [reply irohad/main/impl/pending_transaction_storage_init.cpp:18]: https://imgur.com/3l4KntA.png

File: irohad/main/impl/pending_transaction_storage_init.cpp [hunk 18]
```diff
@@ -15,36 +15,37 @@
...
                                           updated_batches.get_subscriber());
   mst_processor.onPreparedBatches().subscribe(pending_storage_lifetime,
                                               prepared_batch.get_subscriber());
   mst_processor.onExpiredBatches().subscribe(pending_storage_lifetime,
                                              expired_batch.get_subscriber());
-}
+}*/
 
-void PendingTransactionStorageInit::setFinalizedTxsSubscription(
+/*void PendingTransactionStorageInit::setFinalizedTxsSubscription(
```
[the remark this answers is not in the document]
Comment (iceseer) [reply irohad/main/impl/pending_transaction_storage_init.cpp:43]: https://imgur.com/3l4KntA.png

File: irohad/main/application.hpp (hunk 7 above)
Comment (iceseer) [reply irohad/main/application.hpp:385]: https://imgur.com/3l4KntA.png

File: irohad/torii/processor/transaction_processor_impl.hpp (hunk 8 above)
Comment (iceseer) [reply irohad/torii/processor/transaction_processor_impl.hpp:76]: https://imgur.com/3l4KntA.png

File: irohad/consensus/yac/yac.hpp (hunk 9 above)
Comment (iceseer) [reply irohad/consensus/yac/yac.hpp:23]: https://imgur.com/3l4KntA.png

File: irohad/multi_sig_transactions/impl/mst_processor_impl.cpp (hunk 10 above)
Comment (iceseer) [reply irohad/multi_sig_transactions/impl/mst_processor_impl.cpp:13]: https://imgur.com/3l4KntA.png

File: irohad/multi_sig_transactions/storage/mst_storage_impl.hpp (hunk 11 above)
Comment (iceseer) [reply irohad/multi_sig_transactions/storage/mst_storage_impl.hpp:12]: https://imgur.com/3l4KntA.png

File: test/framework/test_subscriber.hpp (hunk 12 above)
Comment (iceseer) [reply test/framework/test_subscriber.hpp:11]: https://imgur.com/3l4KntA.png

File: irohad/ordering/impl/on_demand_connection_manager.hpp (hunk 13 above)
Comment (iceseer) [reply irohad/ordering/impl/on_demand_connection_manager.hpp:14]: https://imgur.com/3l4KntA.png

File: irohad/pending_txs_storage/impl/pending_txs_storage_impl.hpp (hunk 14 above)
Comment (iceseer) [reply irohad/pending_txs_storage/impl/pending_txs_storage_impl.hpp:20]: https://imgur.com/3l4KntA.png

File: irohad/main/application.cpp (hunk 15 above)
Comment (iceseer) [reply irohad/main/application.cpp:133]: https://imgur.com/3l4KntA.png

File: test/framework/test_subscriber.hpp (hunk 12 above)
Comment (iceseer) [reply test/framework/test_subscriber.hpp:12]: ```make_test_subscriber``` uses ```TestSubscriber```, that uses ```rxcpp```.

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 19]
```diff
@@ -51,59 +51,72 @@ namespace iroha {
...
+            hash_gate_(std::move(hash_gate)),
+            delay_func_(std::move(delay_func)) {}
+
+      void YacGateImpl::initialize() {
+        block_creator_subscription_ =
+            iroha::SubscriberCreator<bool, simulator::BlockCreatorEvent>::
+                template create<EventTypes::kOnBlockCreatorEvent,
+                                SubscriptionEngineHandlers::kYac>(
+                    [wptr(weak_from_this())](auto &, auto event) {
+                      if (auto ptr = wptr.lock())
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp:67]: throw here! do not hide from doing the job! Any problem **MUST** be reported.

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 20]
```diff
@@ -51,59 +51,72 @@ namespace iroha {
...
+                        ptr->vote(event);
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp:68]:
```suggestion
                      std::shared_ptr(wptr)->vote(event);
```

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 21]
```diff
@@ -51,59 +51,72 @@ namespace iroha {
...
+              // check ptr ref remains 1
+              if (auto ptr = wptr.lock())
+                visit_in_place(
+                    message,
+                    [&](const CommitMessage &msg) { ptr->handleCommit(msg); },
+                    [&](const RejectMessage &msg) { ptr->handleReject(msg); },
+                    [&](const FutureMessage &msg) { ptr->handleFuture(msg); });
+            });
+
+        outcome_subscription_ =
```
Comment (kuvaldini) [inline irohad/consensus/yac/impl/yac_gate_impl.cpp:85]: Language logic error. You create a Subscriber. But result variable is subscription. Please select what is result subscriber or subscription. subscription is result of subscriber's subscribe().
*Let's be clear in human language and code will also be more clear*

**Pull request closed without merging**